### PR TITLE
Some State Serialization Refactoring

### PR DESCRIPTION
**Pull request by pmuetschard**
In order to share some code between GLES and Vulkan for state serialization.

### CODE REVIEW

File: gapii/cc/spy.cpp [hunk 1]
```diff
@@ -450,58 +449,17 @@ void Spy::onPreStartOfFrame(CallObserver* observer, uint8_t api) {
...
-                observer->encode(&observation);
-            }
-        });
-        observer->encode(this->VulkanSpy::mState);
-        observer->on_slice_encoded(nullptr);
+    saveInitialStateForApi<GlesSpy>("gles-initial-state");
+    saveInitialStateForApi<VulkanSpy>("vulkan-initial-state");
+}
+
+template<typename T>
```
Comment (AWoloszyn) [inline gapii/cc/spy.cpp]: Good use of template!

File: gapis/api/gles/api/egl.api [hunk 2]
```diff
@@ -187,7 +187,12 @@ cmd EGLSurface eglCreateWindowSurface(EGLDisplay          display,
 
 @no_replay
 cmd EGLBoolean eglDestroyContext(EGLDisplay display, EGLContext context) {
-  return ? // TODO
+  if (context == null) {
```
Comment (AWoloszyn) [inline gapis/api/gles/api/egl.api]: `if (context != null) && context in EGLContexts {` ?

**Comment by pmuetschard** [comment]
Tested bloom on Android.

File: gapis/api/gles/api/egl.api (hunk 2 above)
Comment (ben-clayton) [reply gapis/api/gles/api/egl.api]: Agree, this is a bit weird.

Comment (pmuetschard) [reply gapis/api/gles/api/egl.api]: Ooops, sorry about that. Will send a fix.

**Pull request merged**